witness/golang/omniwitness/usbarmory/storage/slots/journal.go

```diff
@@ -0,0 +1,258 @@
+	// DataLen is the length in bytes of the application data associated with this entry.
+	DataLen uint64
+	// DataSHA256 is the SHA256 hash of the application data associated with this entry.
+	DataSHA256 [32]byte
+	// Data is the application data associated with this entry.
+	Data []byte
+}
+
+// entryHeaderSize is the on-disk size of an entry without application data.
+const entryHeaderSize = 4 + 4 + 8 + 32
+
+// Size returns the number of bytes used by this entry record.
+func (e *entry) Size() int {
+	return entryHeaderSize + len(e.Data)
+}
+
+// OpenJournal returns a new journal structure for interacting with a journal stored in the
+// [start, start+length) range of blocks accessible via dev.
+func OpenJournal(dev BlockReaderWriter, start, length uint) (*Journal, error) {
+	j := &Journal{
+		dev:           dev,
+		start:         start,
+		length:        length,
+		maxWriteBytes: (length*dev.BlockSize())/3 - entryHeaderSize,
+	}
+
+	if err := j.init(); err != nil {
+		return nil, err
+	}
+
+	return j, nil
+}
+
+// Data returns the application data from the most recent valid entry in the
+// journal, along with the entry's revision number.
+func (j *Journal) Data() ([]byte, uint32) {
+	return j.current.Data, j.current.Revision
+}
+
+// Update creates a new entry record in the journal for the provided data.
+// The new record's revision will be one greater than the previous record (or 1
+// if no previous record exists).
+func (j *Journal) Update(data []byte) error {
+	if l := len(data); l > int(j.maxWriteBytes) {
+		return fmt.Errorf("attemping to write %d bytes, larger than the max permitted in this journal (%d bytes)", l, j.maxWriteBytes)
+	}
+	e := &entry{
+		Magic:      [4]byte{magic0[0], magic0[1], magic0[2], magic0[3]},
+		Revision:   j.current.Revision + 1,
+		DataLen:    uint64(len(data)),
+		DataSHA256: sha256.Sum256(data),
+		Data:       data,
+	}
+
+	cap := (j.length - j.nextBlock) * j.dev.BlockSize()
+	if cap < uint(e.Size()) {
+		// The record won't fit in the remaining space, so wrap around and write at the beginning.
+		j.nextBlock = j.start
+	}
+	// TODO(al): consider making this more "streamy".
+	buf := &bytes.Buffer{}
+	if err := marshalEntry(e, buf); err != nil {
+		return fmt.Errorf("failed to marshal entry: %v", err)
+	}
+	return j.dev.WriteBlocks(j.nextBlock, padToBlock(buf.Bytes(), j.dev.BlockSize()))
+}
+
+// Init scans the journal to figure out the latest valid record, if any.
+func (j *Journal) init() error {
+	j.nextBlock = j.start
+	// Start where all good stories do: at the beginning!
+	lba := j.start
+	var lastEntry *entry
+	for lba < j.start+j.length {
+		br := newBlockReader(j.dev, lba)
+		e, err := unmarshalEntry(br)
+		if err != nil {
+			if lastEntry != nil {
+				// We already found the lastet record in the journal, so we're done.
+				j.current = *lastEntry
+				j.nextBlock = lba
+				return nil
+			}
+			// Since we couldn't unmarshal the entry struct we'll have to fall back to
+			// scanning all blocks.
```

Comment:
It might be worth extending this comment to be clear about the case this captures. It looks quite generic, but I believe it's quite a special case actually: on the last journal write, it failed during the write, and this write was a wraparound that started at the beginning. In all other cases, either a) there is no data to read (because nothing was ever written), or b) `lastEntry != nil`.

While reading this I went "yeah that makes sense", then "hold on, no", "oh yeah, in case any data got corrupted at rest", "wait a sec, we don't tolerate that. This can never happen" *lightbulb moment* "right, the first element but only after a wraparound".

Reply:
Thanks - I've expanded the comment, does that capture it better?